common/statistics/src/clients/mod.rs

```diff
@@ -0,0 +1,149 @@
+pub enum ClientStatsEvents {
+    /// Packet count events
+    PacketStatistics(packet_statistics::PacketStatisticsEvent),
+    /// Gateway Connection events
+    GatewayConn(gateway_conn_statistics::GatewayStatsEvent),
+    /// Nym API connection events
+    NymApi(nym_api_statistics::NymApiStatsEvent),
+}
+
+/// Controls stats event handling and reporting
+pub struct ClientStatsController {
+    //static infos
+    last_update_time: OffsetDateTime,
+    client_id: String,
+    client_type: String,
+    os_information: os_info::Info,
+
+    // stats collection modules
+    packet_stats: packet_statistics::PacketStatisticsControl,
+    gateway_conn_stats: gateway_conn_statistics::GatewayStatsControl,
+    nym_api_stats: nym_api_statistics::NymApiStatsControl,
+}
+
+impl ClientStatsController {
+    /// Creates a ClientStatsController given a client_id
+    pub fn new(client_id: String, client_type: String) -> Self {
+        ClientStatsController {
+            //Safety : 0 is always a valid number of seconds
+            #[allow(clippy::unwrap_used)]
+            last_update_time: OffsetDateTime::now_utc().replace_second(0).unwrap(), // allow a bigger anonymity set wrt
+            client_id,
+            client_type,
+            os_information: os_info::get(),
+            packet_stats: Default::default(),
+            gateway_conn_stats: Default::default(),
+            nym_api_stats: Default::default(),
+        }
+    }
+    /// Returns a static ClientStatsReport that can be sent somewhere
+    pub fn build_report(&self) -> ClientStatsReport {
+        ClientStatsReport {
+            last_update_time: self.last_update_time,
+            client_id: self.client_id.clone(),
+            client_type: self.client_type.clone(),
+            os_information: self.os_information.clone(),
+            packet_stats: self.packet_stats.report(),
+            gateway_conn_stats: self.gateway_conn_stats.report(),
+            nym_api_stats: self.nym_api_stats.report(),
+        }
+    }
+
+    /// Handle and dispatch incoming stats event
+    pub fn handle_event(&mut self, stats_event: ClientStatsEvents) {
+        match stats_event {
+            ClientStatsEvents::PacketStatistics(event) => self.packet_stats.handle_event(event),
+            ClientStatsEvents::GatewayConn(event) => self.gateway_conn_stats.handle_event(event),
+            ClientStatsEvents::NymApi(event) => self.nym_api_stats.handle_event(event),
+        }
+    }
+
+    /// Reset the metrics to their initial state.
+    ///
+    /// Used to periodically reset the metrics in accordance with periodic reporting strategy
+    pub fn reset(&mut self) {
+        self.nym_api_stats = Default::default();
+        self.gateway_conn_stats = Default::default();
+        //no periodic reset for packet stats
+
+        #[allow(clippy::unwrap_used)] //Safety : 0 is always a valid number of seconds
+        let now = OffsetDateTime::now_utc().replace_second(0).unwrap();
```

Comment:
given you're doing that replacement in multiple places, why not extract it to some function for easier updates?

Reply:
Changed in https://github.com/nymtech/nym/pull/5107/commits/5b263148ffbb19b14d690537e2f622faebadb66a